### PR TITLE
Features/hyperparameter search (but actually includes motor position optimization)

### PR DESCRIPTION
**Pull request by MarleneBusch**
# Description
This PR mainly includes the motor positions optimization as a new feature but also updates the surface reconstruction.
- I have added two actuator parameters to the artist scenario but also to the .h5 files.
- There is a new file with loss functions that can be used interchangeably in the optimizers.
- I have removed code duplicates in the rigid body kinematic module.
- Most of the data files like the scenario.h5 and test data had to be updated after the above changes.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by kalebphipps — CHANGES_REQUESTED** [review]
Thanks for all the changes!

I think everything is looking good in general my comments were mainly regarding two general points:
- Consistency: It seems that some of the new changes (i.e. using a callable loss) are not yet in all the optimizers. Maybe this doesn't make sense, but it would be nice for all optimizers to have a similar structure - i.e. if you call the ``optimize`` function, they expect similar parameters - this would make it easier to perform different tasks across ARTIST. Does it maybe make sense to have an abstract parent `Optimizer` class to ensure the same signature or is this not viable. Also make sure early stopping is implemented everywhere - basically, the functions behave the same everywhere.
- Usability: For some functions there are now a lot of weights and options - most noticeably the surface reconstructor. I have made some suggestions to try and make this a little simpler and remove some parameters - but again, perhaps this is a bad idea.

Let me know what you think and if you have any questions we can discuss them!

File: artist/core/loss_functions.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    Parameters
+    ----------
+    predictions : torch.Tensor
+        The flux distribution.
+        Tensor of shape [1, bitmap_resolution_e, bitmap_resolution_u].
+    targets : torch.Tensor
+        The target distributions.
+        Tensor of shape [number_of_flux_distributions, bitmap_resolution_e, bitmap_resolution_u].
+    **kwargs : Any
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/loss_functions.py:147]: Turns out we dont. I deleted them. I originally included kwargs because I thought mypy needs them.

File: artist/core/loss_functions.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+
+def distribution_loss_kl_divergence(
+    predictions: torch.Tensor, targets: torch.Tensor, **kwargs: Any
+) -> torch.Tensor:
+    """
+    Compute the loss for an optimization using distributions as target.
+
+    The computation is performed elementwise over the last two dimensions and
+    summed to give the kl-divergence as loss. The elements in predictions and targets
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/loss_functions.py:135]: done

File: artist/core/motor_position_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        if rank == 0:
+            log.info("Start the motor positions optimization.")
+
+        for heliostat_group_index in self.ddp_setup["groups_to_ranks_mapping"][rank]:
+            heliostat_group = self.scenario.heliostat_field.heliostat_groups[
+                heliostat_group_index
+            ]
+
+            # Align all heliostats once, to the given incident ray direction and target, to set initial
+            # motor positions. The motor positions are set automatically within the align from incident ray method.
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/motor_position_optimizer.py:146]: done

File: tests/core/test_surface_reconstructor.py [hunk 4]
```diff
@@ -66,35 +84,168 @@ def test_surface_reconstructor(
...
-            atol=5e-2,
-            rtol=5e-2,
+            heliostat_group.nurbs_control_points,
+            expected,
+            atol=5e-5,
+            rtol=5e-5,
         )
+
+
+def test_fixate_control_points_on_outer_edges(
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply tests/core/test_surface_reconstructor.py:121]: done

File: artist/core/motor_position_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+            # optimization of this heliostat. This mismatch makes it impossible to choose a single learning
+            # rate that works robustly across all heliostats.
+            # The reparametrization of the optimizable parameter (motor positions) defines the optimizable
+            # parameter as:
+            # motor_positions_optimized = tanh(torch.nn.Parameter(optimizable_parameter))
+            # The true motor positions can be reconstructed by:
+            # motor_positions = initial_motor_positions + motor_positions_normalized * scale
+            # where scale defines the range (e.g. up to ~80,000) for adjustments.
+            # By optimizing as explained above instead of raw motor positions, every heliostat sees updates
+            # of comparable relative magnitude, regardless of the absolute size of its motors positions.
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/motor_position_optimizer.py:189]: I have moved it to the docstring and added the math notation. But there isnt rally a reference for this, i think, its just a reparametrization.

File: artist/core/motor_position_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        self.optimization_goal = optimization_goal
+        self.bitmap_resolution = bitmap_resolution.to(device)
+        self.initial_learning_rate = initial_learning_rate
+        self.tolerance = tolerance
+        self.max_epoch = max_epoch
+        self.num_log = num_log
+
+    def optimize(
+        self,
+        loss_function: Callable[..., torch.Tensor],
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/motor_position_optimizer.py:117]: All optimze methods from the 3 optimizers (kinematic, surface, motor pos) now take the loss_function and device as parameters. They all return a loss tensor.
Do you want me to change the names of all optimize methods as well?
For the kinematic optimzation it is: 
`_ = kinematic_optimizer.optimize(loss_function=loss_function, device=device)`
For the surface reconstructor it is: 
`_ = surface_reconstructor.reconstruct_surfaces(loss_function=loss_function, device=device)`
For motor positions optimizer it is: 
`_ = motor_positions_optimizer.optimize(loss_function=loss_function, device=device)`

I am wondering should i rename `reconstruct_surfaces()` to `optimize()`
and the class `KinematicOptimizer` to `KinematicCalibrator`?

[the remark this answers is not in the document]
Comment (kalebphipps) [reply artist/core/motor_position_optimizer.py:117]: I think the functions and classes should have names that are intuitive for people working in the field. In this case I think ``surface_reconstructor.reconstruct_surface`` makes sense, since we are not "optimising" the surface but learning it to create an accurate digital twin.

For the kinematic I am not sure what the process is - I think either as it is, or to have a ``KinematicCalibrator`` with the function ``calibrate`` instead of ``optimize``. What makes more intuitive sense for you?

File: artist/core/kinematic_optimizer.py [hunk 7]
```diff
@@ -304,74 +268,145 @@ def _optimize_kinematic_parameters_with_raytracing(
...
+                        target_widths=self.scenario.target_areas.dimensions[
+                            target_area_mask
+                        ][:, 0],
+                        target_heights=self.scenario.target_areas.dimensions[
+                            target_area_mask
+                        ][:, 1],
+                        device=device,
+                    )
+
+                    loss = (focal_spots - focal_spots_measured).abs().mean()
```
[the remark this answers is not in the document]
Comment (kalebphipps) [reply artist/core/kinematic_optimizer.py:372]: Having the functionality (even if overkill) can't really hurt I don''t think. So I'm happy for you to resolve this conversation!

File: artist/core/surface_reconstructor.py [hunk 8]
```diff
@@ -109,202 +145,601 @@ def __init__(
...
+        self.sigma_tv_loss = sigma_tv_loss
+        self.early_stopping_threshold = early_stopping_threshold
         self.initial_learning_rate = initial_learning_rate
         self.tolerance = tolerance
         self.max_epoch = max_epoch
+        self.num_log = num_log
 
     def reconstruct_surfaces(
         self,
+        loss_function: Callable[..., torch.Tensor],
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/surface_reconstructor.py:173]: i have done this as best as i could, but it was difficult and i am not sure if i chose a good way to implement this. Looking forward to your review.

File: artist/core/loss_functions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/loss_functions.py:147]: if you know another way to avoid this propblem, let me know.

File: examples/set_up.py
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply examples/set_up.py:1]: done

File: examples/__init__.py
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply examples/__init__.py:1]: done

File: artist/core/loss_functions.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+
+        Parameters
+        ----------
+        prediction : torch.Tensor
+            The predicted values.
+            Tensor of shape [number_of_samples, bitmap_resolution_e, bitmap_resolution_u].
+        ground_truth : torch.Tensor
+            The ground truth.
+            Tensor of shape [number_of_samples, 4].
+        **kwargs : Any
```
Comment (kalebphipps) [inline artist/core/loss_functions.py:148]: In cases where specific key word arguments are expected, i.e. here I would include these in the doctoring.

i.e.

```
Keyword arguments.
The ``target_area_mask`` and optionally ``device`` are expected for the focal spot loss.
```

File: artist/core/loss_functions.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+            Keyword arguments.
+
+        Returns
+        -------
+        torch.Tensor
+            The summed MSE focal spot loss reduced along the specified dimensions.
+            Tensor of shape [number_of_samples].
+        """
+        device = get_device(device=kwargs["device"])
+
```
Comment (kalebphipps) [inline artist/core/loss_functions.py:158]: Should we raise an error if unexpected keyword arguments are required, i.e.

```
# Define expected keyword arguments for clarity and validation.
        expected_kwargs = {"target_area_mask", "device"}
        for key in kwargs:
            if key not in expected_kwargs:
                raise ValueError("The Focal spot loss only expects .... but you provided {key}. Please only provide the expected keyword arguments when using ...")
```

File: scenario.h5
Comment (kalebphipps) [inline scenario.h5:1]: This is a random scenario file in the main Git folder - was this committed by mistake? I think it should be removed!

File: artist/core/loss_functions.py (hunk 10 above)
Comment (kalebphipps) [reply artist/core/loss_functions.py:158]: This should work at the moment since these are the only extra parameters that could be passed right? If not, i.e. if there is a good reason for not doing this then we can leave it out :)
